### PR TITLE
feat: add union types

### LINKED ISSUE #24: Add support for union types
A union type is a set consisting of multiple non-overlapping types. With union types, we can have function signatures of the type:

```elixir
fn foo : {:ok, Int} | {:error, String} do
  if true do
    {:ok, 123}
  else
    {:error, "Something went wrong"}
  end
end
```

We'll need to understand the concept of subtypes before we start working on this. [This wiki](https://en.wikipedia.org/wiki/Subtyping) is a good introduction.

In this example:
image
duck is a subtype of bird (which is its supertype), which means it can be substituted wherever a bird is expected. Similarly, a union type is a supertype of each member of the set.

Changes required:

1. Allow if-else branches to return different types. The inferred type of the if-else expression will be a union of the two types. This can be represented using a new "Union" struct that can be defined [here](https://github.com/fika-lang/fika/blob/master/lib/fika/types.ex) like the FunctionRef struct. The to_string function of the struct will join the types separated by " | ".
2. When [inferring the return type of a function](https://github.com/fika-lang/fika/blob/master/lib/fika/type_checker.ex#L52), we'll now expand union types and add all possible function types in [add_function_type](https://github.com/fika-lang/fika/blob/master/lib/fika/type_checker.ex#L402). Example:

```elixir
# A function that takes x which can be either :foo or :bar and returns either :ok or :error
fn foo(x: :foo | :bar) : :ok | :error do
  # ...
end
```

The type checker will add the following function types to its "function_types" map in env:
1. module.foo(:foo) => :ok 
2. module.foo(:foo) => :error
3. module.foo(:bar) => :ok
4. module.foo(:bar) => :error 

We can optimize the storing and checking of function types later, but this should do for now.

Please let me know if I've missed something here!

### PR DESCRIPTION
**Pull request by polvalente**
This PR aims tackles union types

closes #24

### CODE REVIEW

**Comment by polvalente** [comment]
@emilsoman here's the first draft. Does this look good to you, or are there any major structural changes you want?

**Comment by emilsoman** [comment]
I noticed that we return a list of types instead of a string from the parser after parsing types. I am guessing it's because you wanted to check if the return type of a function is one of the possible types in the type spec of the function? I think that check should change to an "exact match" (that means we can still use strings). Reason - even though technically a function returning `:foo` can fit into a slot for `:foo | :bar`, the return type in the signature should be accurate about what the function actually returns. For example, this should not be allowed:

```elixir
fn sum(a: Int, b: Int) : Int | String do
  a + b
end
```
because the function never returns a string and the signature is lying to the developer who will now have to conditionally handle ints and strings wherever this function is used. The compiler should report a type checking error saying the return type is specified as `Int | String`, but the function actually returns `Int`. This also means the parser can just return a string instead of a list of types. I agree that we should change this to a richer data structure later for better comparisons and improving our error messages, but this is not required for now.

Also, I am guessing the type checker changes are still WIP? A few things to note here:

1. if-else can now return two different types from their two branches and the type of the if-else expression will be the union type of these branches.

2. When passing args to functions, we need new checks now, examples:

```elixir
# A function foo
#  that takes an arg x which can be an Int or a String
#  and returns one of Int, String, Bool
fn foo(x: Int | String) : Int | String | Bool

# A function bar
#  which takes an arg x which is a function ref
#    and the function it references should be able to take an Int and it should return a Bool.
fn bar(x : Fn(Int -> Bool)) : ...
```

This means`bar(&foo(Int))` should not be allowed because when foo is given an Int, it may return one of Int, String or Bool, but here, the function is expected to return only a Bool. But if bar was defined like `fn bar(x: Fn(Int -> Int | String | Bool))`, then we should be able to call `bar(&foo(Int))`.

Also, another case:

```elixir
# x can be either an Int or a String depending on the condition evaluation during runtime.
x =
  if cond do
    123
  else
    "Hello world"
  end

# Given two function with signatures below:
fn func1(x: Int) : Int
fn func2(x: Int | String | Bool) : Int

func1(x) # not allowed
func2(x) # allowed
```

Hope it makes sense and you get where I'm going with this.

**Comment by emilsoman** [comment]
Also, I've simplified the parser a bit (removed a few unnecessary Enum.joins and exporting the atom parser) and added the union type parser here : https://github.com/fika-lang/fika/compare/union_type_parser?diff=split&expand=1 please take these changes into your branch if it makes sense

**Comment by polvalente** [comment]
> I noticed that we return a list of types instead of a string from the parser after parsing types. I am guessing it's because you wanted to check if the return type of a function is one of the possible types in the type spec of the function? I think that check should change to an "exact match" (that means we can still use strings). Reason - even though technically a function returning :foo can fit into a slot for :foo | :bar, the return type in the signature should be accurate about what the function actually returns. For example, this should not be allowed:

Ok, so basically that `Enum.any?` must change into an `Enum.all?`. However, I still think we need to parse types as lists so we can iterate through all of them individually. Otherwise, I'm gonna have to split the list anyway to check each type, or if not parsing, then some function that is inferred as ":foo | :bar" will not match exactly ":bar | :foo" whereas they are the same thing. Whad do you think about this?

> Also, I am guessing the type checker changes are still WIP? A few things to note here:

Yeah, I've only changed the parser, basically, and made the other tests pass.

**Comment by emilsoman** [comment]
> some function that is inferred as ":foo | :bar" will not match exactly ":bar | :foo" whereas they are the same thing

You are absolutely right, this is the part I was talking about in this line : 
>  I agree that we should change this to a richer data structure later for better comparisons and improving our error messages

Let's see what happens without it:

```elixir
fn fun : :bar | :foo do
  ...

  if cond do
    fn_that_returns_foo()
  else
    fn_that_returns_bar()
  end
end 
```

TypeChecker reports the error: "Expected `:bar | :foo`, but got `:foo | :bar`. Now is it something we can live with for now? I thought about it and I feel this error message is okay, because the change I'd need to make to fix this is to make the signature reflect the order in which the return types are accumulated which is not so bad (some may even think this is a feature that helps readability of said function). But then again, I think enforcing order in union types isn't the right thing to do and why I suggested doing that is only to limit this PR's scope and introduce struct comparisons for type checking in a different PR. It's fine to use lists as intermediate step before that too, so I agree with your suggestion!

File: lib/fika/cli.ex [hunk 1]
```diff
@@ -46,7 +46,7 @@ defmodule Fika.Cli do
     try do
       Logger.debug("Calling :#{module}.#{function}")
       {result, _binding} = Code.eval_string(~s':"#{module}".#{function}')
-      IO.inspect(result)
+      result |> inspect() |> IO.puts()
```
Comment (emilsoman) [inline lib/fika/cli.ex:49]: Is this change required?

File: lib/fika/parser/helper.ex [hunk 2]
```diff
@@ -147,6 +147,13 @@ defmodule Fika.Parser.Helper do
     {:atom, line, value}
   end
 
+  def join(enumerable, separator \\ "") do
+    Enum.reduce(enumerable, "", fn
+      {:atom, _, value}, acc -> acc <> separator <> ":#{value}"
+      value, acc -> acc <> separator <> value
+    end)
```
Comment (emilsoman) [inline lib/fika/parser/helper.ex:141]: In my branch, I wrote a parser for atom types that removes the need for this conditional treatment of atoms here. Maybe we can use that?

File: lib/fika/env.ex [hunk 3]
```diff
@@ -10,25 +10,25 @@ defmodule Fika.Env do
...
+        "kernel.*(Int,Float)" => ["Float"],
+        "kernel.*(Float,Int)" => ["Float"],
+        "kernel.*(Float,Float)" => ["Float"],
+        "kernel./(Int,Int)" => ["Float"],
+        "kernel./(Int,Float)" => ["Float"],
+        "kernel./(Float,Int)" => ["Float"],
+        "kernel./(Float,Float)" => ["Float"],
+        "kernel.|(Bool,Bool)" => ["Bool"],
+        "kernel.&(Bool,Bool)" => ["Bool"],
+        "kernel.!(Bool)" => ["Bool"]
```
Comment (emilsoman) [inline lib/fika/env.ex]: It would be nice if we can use lists only for union types. Let's discuss this in discord if needed.

**Review by emilsoman — COMMENTED** [review]
As you said, let's tackle this in multiple PRs if the scope is getting too big. I will review the rest of the PR after we have finalized the parser changes.

File: lib/fika/cli.ex (hunk 1 above)
Comment (polvalente) [reply lib/fika/cli.ex:49]: Not really, but since this is an intentional print, it makes debuggin `IO.inspect`s easier to find when commiting

File: lib/fika/parser/helper.ex (hunk 2 above)
Comment (polvalente) [reply lib/fika/parser/helper.ex:141]: Yeah! I'll try to incorporate that here!

File: lib/fika/env.ex (hunk 3 above)
Comment (polvalente) [reply lib/fika/env.ex]: I agree. My plan of action is:

- Incorporate the new atom parser
- Remove the lists when the type is not an union

File: lib/fika/parser/helper.ex (hunk 2 above)
Comment (polvalente) [reply lib/fika/parser/helper.ex:141]: I still need a special mapping for atoms here, but only because we need to render the ":" when converting to string :)

Otherwise, atoms are atoms as your branch intended!

File: lib/fika/parser/types.ex [hunk 4]
```diff
@@ -89,16 +89,31 @@ defmodule Fika.Parser.Types do
     |> reduce({Enum, :join, []})
     |> label("tuple type")
 
-  type =
+  atom =
+    ignore(string(":"))
+    |> concat(identifier_str)
+    |> reduce({Helper, :to_atom, []})
```
Comment (polvalente) [inline lib/fika/parser/types.ex:95]: @emilsoman This combinator was returning `[":", "value"]` instead of `":value"`. Do you see any way to avoid needing this `to_atom` call?

**Comment by polvalente** [comment]
closed in favor of #43
